File: Client/Assets/Scripts/Cards Scripts/Card.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;


public interface ICommand {
    void execute(object param);
}

public class Card {

    private const string characterType = "Personagem";

    private int id, cost, overcost = 0, cooldown = 1, actual_cooldown;
    private int? attack, life;
    public int original_cost;   

    public Effects onActivateEffect;
    
    private string name, type;
    
    private int actual_damage, actual_life, siege_dmg;


    private bool immuny, invisible, can_act, ranged, charge;

    private Position position;

    private int player_controller;


    public Card (int id, string name, string type, int cost, int damage, int life, int cooldown, int overcost) {
        this.id = id;
        this.cost = cost;
        this.original_cost = cost;
        this.actual_cooldown = 0;
        this.name = name;

        this.cooldown = cooldown;
        this.overcost = overcost;

        this.attack = damage;
        this.life = life;
        this.immuny = false;
        this.invisible = false;
        this.actual_damage = this.attack.GetValueOrDefault();
        this.actual_life = this.life.GetValueOrDefault();
        if (this.attack > 0)
            this.siege_dmg = 1;
        else
            this.siege_dmg = 0;

        this.type = type;
     
    }

    public void setPlayerId(int player_id) {
        this.player_controller = player_id;
    }

    public bool isMagic() {
        return this.type == "Magia";
    }

    public void setSiegeDmg(int siege_dmg) {
        this.siege_dmg = siege_dmg;
    }

    public bool belongsToPlayer(int player_id) {
        return player_id == this.player_controller;
    }

    public void setCharge(bool charge) {
        this.charge = charge;
    }

    public bool hasCharge() {
        return this.charge;
    }

    /// <summary>
    /// Efeito de carta que acontece quando uma criatura entra.
    /// </summary>
    public void onPlayEnter() {
        if (this.hasCharge())
            this.can_act = true;
    }


    public bool canAct() {
        return can_act;
    }
    public Position getPosition() {
        return position;
    }

    public void setRanged(bool range) {
        this.ranged = true;
    }

    public int getPlayerId() {
        return this.player_controller;
    }
	
    public bool isCreature() {
        return this.type == characterType;
    }

    public void resetCard() {
        this.resetStatus();
        this.increaseCost();
    }

    
    /// <summary>
    /// Função chamada se a carta se encontra na mão do player.
    /// </summary>
    public void endOfTurnInHand() {
        if (this.actual_cooldown > 0) {
            this.actual_cooldown--; 
        }
    }

    public int getActualSiegeDmg() {
        return this.siege_dmg;
    }

    public bool canPlay() {
        return this.actual_cooldown == 0;
    }

    private void resetStatus() {
        if (isCreature()) {
            this.actual_damage = (int)this.attack;
            this.actual_life = (int)this.life;
            this.can_act = false;
        }
        this.actual_cooldown = this.cooldown;
    }

    public void increaseCost() {
        this.cost += this.overcost;
    }

    public void setPosition(Position position) {
        this.position = position;
    }

    public void setCanAct (bool can_act) {
        this.can_act = can_act;
    }

    public int getSiegeDmg() {
        return siege_dmg;
    }

    public int getID(){
        return id;
    }

    public bool isDead() {
        return this.actual_life <= 0;
    }

    public bool isRange() {
        return ranged;
    }

    public bool isImmuny() {
        return immuny;
    }

    public bool isInvisible() {
        return invisible;
    }

    public int? getAttack() {
        return this.attack;
    }

    public int? getLife() {
        return this.life;
    }

    public int getCooldown() {
        return this.cooldown;
    }

    public int getActualCooldown() {
        return this.actual_cooldown;
    }

    public int getActualAttack() {
        return actual_damage;
    }

    public int getActualLife() {
        return actual_life;
    }

    public void changeActualLife(int life) {
        actual_life += life;
        if (actual_life > this.life)
            actual_life = life;
    }

    public void dealDamage(int dmg) {
        this.actual_life -= dmg;
    }

    public int getCost() {
        return cost;
    }

}

File: Server/src/javaapplication1/Card.java

package javaapplication1;

/**
 *
 * @author Pedro
 */
public class Card {
    
    ICommand onPlayEffect;
    ICommand onDeathEffect;
    
    private Game game_instance;

    private int id;
    private int attack, life, cost;
    
    private int actual_attack, actual_life, actual_siegedmg;
    
    private boolean able_to_attack, ranged;
    
    private String name, type, text;

    public Card (int id, String name, String type,  int cost, int attack, int life, int cooldown, int overcost){
        this.id = id;
        this.type = type;
        this.name = name;
        this.cooldown = 1;
        this.attack = attack;
        this.life = life;
        this.actual_attack = attack;
        this.actual_cooldown = 0;
        this.cost = cost;
        this.actual_life = life;
        this.actual_siegedmg = 1;
        this.ranged = false;
        this.able_to_attack = false;
    }
    
    
    
    public void setGame(Game game){
        this.game_instance = game;
    }
    
    
    public int getAttack() {
        return attack;
    }

    public void setAttack(int attack) {
        this.attack = attack;
    }

    public int getLife() {
        return life;
    }

    public Game getGameInstance(){
        return this.game_instance;
    }
    
    public FieldControl getFieldControl(){
        return this.game_instance.getFieldControl();
    }
    
    public void setLife(int life) {
        this.life = life;
    }
    
    public void takeDamage(int dmg){
        this.actual_life -= dmg;
        if (this.actual_life <= 0){
            if (this.onDeathEffect != null)
                this.onDeathEffect.execute(null);
        }
    }
    
    public boolean isDead(){
        return this.actual_life <= 0;
    }

    public int getCost() {
        return cost;
    }

    public boolean canPlay(){
        return this.actual_cooldown == 0;
    }
    
    public void endTurnInHand(){
        if (this.actual_cooldown > 0)
            this.actual_cooldown -= 1;
    }
    
    
    public void setCost(int cost) {
        this.cost = cost;
    }

    public int getActual_attack() {
        return actual_attack;
    }

    public void setActual_attack(int actual_attack) {
        this.actual_attack = actual_attack;
    }

    public int getActual_life() {
        return actual_life;
    }

    public void setActual_life(int actual_life) {
        this.actual_life = actual_life;
    }

    public int getActual_siegedmg() {
        return actual_siegedmg;
    }

    public void setActual_siegedmg(int actual_siegedmg) {
        this.actual_siegedmg = actual_siegedmg;
    }

    public boolean isAble_to_attack() {
        return able_to_attack;
    }

    public void setAble_to_attack(boolean able_to_attack) {
        this.able_to_attack = able_to_attack;
    }

    public boolean isRanged() {
        return ranged;
    }

    public void setRanged(boolean ranged) {
        this.ranged = ranged;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public int getPlayer_id() {
        return player_id;
    }

    public void setPlayer_id(int player_id) {
        this.player_id = player_id;
    }

    public int getCooldown() {
        return cooldown;
    }

    public void setCooldown(int cooldown) {
        this.cooldown = cooldown;
    }

    public int getActual_cooldown() {
        return actual_cooldown;
    }

    public void setActual_cooldown(int actual_cooldown) {
        this.actual_cooldown = actual_cooldown;
    }
    
    private int player_id, cooldown, actual_cooldown;
    

    
    public boolean canAttack(){
        return this.able_to_attack;
    }
    
    public void setAttack(boolean can_attack){
        this.able_to_attack = can_attack;
    }
    
    public int getPlayerId(){
        return this.player_id;
    }
    
    
    public void onPlayed(){
        if (this.onPlayEffect != null)
            onPlayEffect.execute(null);
    }
    
    
    
    public void reset(){
        this.actual_attack = this.attack;
        this.actual_life = this.life;
        this.able_to_attack = false;
        this.actual_cooldown = this.cooldown;
    }
    
    public int getId(){
        return id;
    }
    
    
}
